lib/wca_live/accounts/scoretaking_token.ex

```diff
@@ -0,0 +1,70 @@
+defmodule WcaLive.Accounts.ScoretakingToken do
+  @moduledoc """
+  A competition-scoped token for scoretaking API calls from external
+  systems.
+  """
+
+  use WcaLive.Schema
+
+  import Ecto.Query
+
+  alias WcaLive.{Accounts, Competitions}
+  alias WcaLive.Accounts.ScoretakingToken
+
+  @rand_size 32
+
+  @validity_in_days 7
+
+  schema "scoretaking_tokens" do
+    field :token, :string
+
+    belongs_to :user, Accounts.User
+    belongs_to :competition, Competitions.Competition
+
+    timestamps(updated_at: false)
+  end
+
+  @doc """
+  Generates a new competition-scoped token for the given user.
+  """
+  @spec build_scoretaking_token(%Accounts.User{}, %Competitions.Competition{}) :: %__MODULE__{}
+  def build_scoretaking_token(user, competition) do
+    token = :crypto.strong_rand_bytes(@rand_size) |> Base.url_encode64(padding: false)
+    %ScoretakingToken{token: token, user_id: user.id, competition_id: competition.id}
+  end
+
+  @doc """
+  Returns query that verifies the token and returns the user and
+  competition, if any.
+
+  The token is valid if it matches the value in the database and it
+  has not expired.
+  """
+  @spec verify_token_query(String.t()) :: Ecto.Query.t()
+  def verify_token_query(token) do
+    from(token in token_query(token),
+      join: user in assoc(token, :user),
+      join: competition in assoc(token, :competition),
+      where: token.inserted_at > ago(@validity_in_days, "day"),
+      select: {user, competition}
+    )
+  end
+
+  @doc """
+  Returns query to lookup token struct for the given token value.
+  """
+  @spec token_query(String.t()) :: Ecto.Query.t()
+  def token_query(token) do
+    from(ScoretakingToken, where: [token: ^token])
```

Comment:
If I'm reading this correctly, it looks like we're storing the API tokens in plaintext. To me, this feels like storing passwords in plaintext (someone with a dump of your database would be able to turn around and make API calls to the live site). Just going off of my personal experience, websites seems to be of 2 minds about this:


- I'm guessing that GitHub hashes their personal access tokens, because they emphasize that you only get to see it once.
![image](https://github.com/thewca/wca-live/assets/277474/cc5d62cc-c0e7-4883-8e72-b09b1ce89a63)
- Datadog definitely stores their API keys in plaintext, you can get back to them anytime:
![image](https://github.com/thewca/wca-live/assets/277474/7d047361-80fd-4f12-85f0-7db5f8455fd3)
- I wasn't able to find a great resource about by Googling (searching for "storing api tokens" turns up a lot of results about people who are writing *clients* for various apis). I did find [this opinionated SO answer](https://security.stackexchange.com/questions/180345/do-i-need-to-hash-or-encrypt-api-keys-before-storing-them-in-a-database#:~:text=Yes%2C%20you%20should%20absolutely%20hash,note%20that's%20hashed%20%2D%20not%20encrypted.) that says you should "absolutely" hash API keys, but it doesn't get into the nuance of the different user experience it requires.


For the record, I'm not suggesting a redesign. I'm just curious how much thought you put into this? If you did, it would be nice to add a comment or update your commit message with thoughts on why you opted to do things this way.

Reply:
Hey, this is a very good call! The difference from passwords is that tokens usually have expiration of some kind (in our case they are competition-specific, so the surface area is pretty small), and also removing all tokens is more of an option. The main usability tradeoff is whether we want to always make the token accessible to the user, and also hashing the token for lookup on every request (but that's not really a big price). All that to say, there is really no reason not to hash, so I will change it :)

Reply:
> The difference from passwords is that tokens usually have expiration of some kind (in our case they are competition-specific, so the surface area is pretty small)

Ah, I didn't realized these were scoped to a competition. I thought these were long-lived secrets that I might bake into my automated scoretaking tool.

> whether we want to always make the token accessible to the user

I agree, this feels like the biggest consideration to me.

> All that to say, there is really no reason not to hash, so I will change it :)

Cool! Sorry to create more work for you.

Reply:
Haha, glad to change it now rather than once it's deployed :D